compressors/rANS.py

```diff
@@ -0,0 +1,335 @@
+        symbol_ind = self.find_bin(cum_prob_list, slot)
+        s = self.freqs.alphabet[symbol_ind]
+
+        # retrieve prev state
+        prev_state = block_id * self.freqs.frequency(s) + slot - self.freqs.cumulative_freq_dict[s]
+        return s, prev_state
+
+    def expand_state(self, state: int, encoded_bitarray: BitArray) -> Tuple[int, int]:
+        # remap the state into the acceptable range
+        num_bits = 0
+        while state < self.params.RANGE_FACTOR * self.freqs.total_freq:
+            state_remainder = bitarray_to_uint(
+                encoded_bitarray[num_bits : num_bits + self.params.NUM_BITS_OUT]
+            )
+            num_bits += self.params.NUM_BITS_OUT
+            state = (state << self.params.NUM_BITS_OUT) + state_remainder
+        return state, num_bits
+
+    def decode_symbol(self, state: int, encoded_bitarray: BitArray):
+        # base rANS decoding step
+        s, state = self.rans_base_decode_step(state)
+
+        # remap the state into the acceptable range
+        state, num_bits_used_by_expand_state = self.expand_state(state, encoded_bitarray)
+        return s, state, num_bits_used_by_expand_state
+
+    def decode_block(self, encoded_bitarray: BitArray):
+        # get data block size
+        data_block_size_bitarray = encoded_bitarray[: self.params.DATA_BLOCK_SIZE_BITS]
+        input_data_block_size = bitarray_to_uint(data_block_size_bitarray)
+        num_bits_consumed = self.params.DATA_BLOCK_SIZE_BITS
+
+        # get the final state
+        num_state_bits = self.params.num_state_bits(self.freqs.total_freq)
+        state = bitarray_to_uint(
+            encoded_bitarray[num_bits_consumed : num_bits_consumed + num_state_bits]
+        )
+        num_bits_consumed += num_state_bits
+
+        # perform the decoding
+        decoded_data_list = []
+        for _ in range(input_data_block_size):
+            s, state, num_symbol_bits = self.decode_symbol(
+                state, encoded_bitarray[num_bits_consumed:]
+            )
+
+            # rANS decoder decodes symbols in the reverse direction,
+            # so we add newly decoded symbol at the beginning
+            decoded_data_list = [s] + decoded_data_list
+            num_bits_consumed += num_symbol_bits
+
+        # Finally, as a sanity check, ensure that the end state should be equal to the initial state
+        assert state == self.params.initial_state(self.freqs)
+
+        return DataBlock(decoded_data_list), num_bits_consumed
+
+
+######################################## TESTS ##########################################
+
+
+def _test_rANS_coding(freq, data_size, seed):
+    """Core testing function for rANS"""
+    prob_dist = freq.get_prob_dist()
+
+    # generate random data
+    data_block = get_random_data_block(prob_dist, data_size, seed=seed)
+
+    # get optimal codelen
+    avg_log_prob = get_mean_log_prob(prob_dist, data_block)
+
+    # create encoder decoder
+    rans_params = rANSParams()
```

Comment:
also test a couple different values of NUM_BITS_OUT and RANGE_FACTOR?

Reply:
done